docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,883 @@
+  thus making it possible for the project to continue to evolve in a way that is predictable for the developers
+  implementing Runtime Extensions.
+
+- As a Cluster API maintainer I would like to have a set of tools, utilities and conventions making it possible
+  to implement new Runtime Hooks quickly and consistently across the code base.
+
+### Implementation Details/Notes/Constraints
+
+The proposed solution is designed with the intent to make developing Runtime Extensions as simple as possible, because
+the success of this feature depends on its speed/rate of adoption in the ecosystem.
+
+Accordingly, the proposed solution relies on a well-known, battle tested integration pattern, RESTful APIs.
+A nice side effect of this choice is the possibility to leverage on a set api-machinery tooling and practices the
+Cluster API maintainers are well-used to.
+
+It is also important to notice that the model based on Runtime Hooks and Runtime Extensions implies two separate
+personas being involved, each one with its own responsibilities in the process:
+
+![overview](images/runtime-sdk/overview.png)
+
+The Runtime SDK rules defined in this document are a critical element of the above split of responsibilities,
+defining expectations for each of the above personas.
+
+#### Cluster API Runtime Hooks vs Kubernetes admission webhooks
+
+Runtime Hooks are inspired by Kubernetes admission webhooks, but there is one key difference that splits them apart:
+
+- Admission webhooks are strictly linked to Kubernetes API Server/etcd **CRUD operations** e.g. Create or Update
+  Cluster in etcd.
+- Runtime Hooks can be used to define **arbitrary operations**, e.g. Cluster.BeforeUpgrade, Machine.Remediate etc.
+
+In other words, Runtime Hooks are not concerned about “low-level” details of how Kubernetes handles objects in the
+API Server/etcd; Runtime Hooks instead focus on “high-level” events of a Cluster’s lifecycle.
+
+Please note that, no matter the similarities in some part of the design, users should not make assumptions about
+Runtime Hooks having properties or behaviors typical of Kubernetes admission webhooks unless they are explicitly
+defined in the following paragraphs.
+
+#### Runtime SDK rules
+
+As this proposal is based on RESTful APIs, we are using [OpenAPI Specification v3.0.0](https://swagger.io/specification/) [1] 
+to document Runtime Hooks supported by Cluster API.
+
+Most specifically, a single OpenAPI document providing specification for all the Runtime Hooks supported by a
+Cluster API release will be added to the release artifacts; users can rely on https://editor.swagger.io/ or similar
+tools to view the specification, and during implementation we will consider adding similar view to the Cluster API
+book as well e.g.
+
+![overview](images/runtime-sdk/swagger-ui.png)
+
+Each Runtime Hook will be defined by one (or more) RESTful APIs implemented as a `POST` operation; each operation
```

Comment:
`POST` are generally not idempotent. Was `PUT` or `PATCH` considered?

Reply:
Good question. 

One data point. Kubernetes is using POST too for Validationg/Mutating webhooks.

Reply:
Maybe Kubernetes is using `POST` because for AdmissionRequests they are not creating the same resource all over again. It's a new one every time they call an admission hook.

I think it's similar for us + that we expect when we create multiple "different" resources we are always getting the same response. But we're not create+updating the same resource (only the content is the same)

Reply:
I will keep this consistent with Kubernetes for now...